Unhardcoded wait timeout for minion refresh event

Oloremo opened the pull request:
### What does this PR do?
Unhardcoded wait timeout for minion refresh event

### What issues does this PR fix or reference?
Partially Fixes: #20590

### Previous Behavior
Timeout was hardcoded to 30 seconds

### New Behavior
Timeout is 30 seconds by default but will be overridden if `timeout` argument was passed in `saltutil.refresh_pillar` for example.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes

Oloremo commented on salt/modules/saltutil.py:
```diff
@@ -378,28 +378,31 @@ def refresh_grains(**kwargs):
     # them, and the module refresh process reloads the grains and assigns the
     # newly-reloaded grains to each execution module's __grains__ dunder.
-    if _refresh_pillar:
-        # we don't need to call refresh_modules here because it's done by refresh_pillar
```
I think that statement was a lie, I checked the refresh_pillar function and it didn't do refresh_modules. 
Please correct me if I'm wrong.

Oloremo committed:
Added an ability to refresh grains synchronously

max-arnold commented:
A similar feature was implemented for pillar: https://github.com/saltstack/salt/pull/56881

whytewolf commented:
@Oloremo thank you for the PR. looking at it does look like it changes the functionality in #56881 so this would be correct to add. 

Can you fix the conflict with the pre-commit-config?

Oloremo commented:
@whytewolf almost forgot re this PR, I didn't have time to write a test so I add WIP and didn't expect it to be noticed.
Let me try to update it.

whytewolf commented:
awesome. when you do get time to add a test all we would need is one that tests that the wait argument is indeed changed when a timeout is supplied.

Oloremo commented:
@whytewolf 
>  when you do get time to add a test 

No idea, I tried to setup dev env on a new MacBook and I failed. It installs half of the internet, some libs are absent I honestly have no idea why it's so heavy just to run black and isort? I guess the docs part is complicated and I'd suggest splitting pre-commit to the necessary and full or something.

whytewolf commented:
> @whytewolf
> 
> > when you do get time to add a test
> 
> No idea, I tried to setup dev env on a new MacBook and I failed. It installs half of the internet, some libs are absent I honestly have no idea why it's so heavy just to run black and isort? I guess the docs part is complicated and I'd suggest splitting pre-commit to the necessary and full or something.

it isn't just black, isort and lint. the heaviest part of pre-commit is actually the requisite checking. which is why it pulls down everything to make sure it knows what gets pulled down. this includes every requisite needed to run tests. which is needed to make sure the test environments are not broken and have every lib needed to run the tests.

Also, currently I don't think salt supports M1 based macs so that might be one of the reasons you might be having trouble.

twangboy committed:
tests: refresh_modules sync timeout coverage; fix conftest psutil patch on old psutil

Exercise synchronous saltutil.refresh_modules by mocking event.fire and get_event:
assert MINION_MOD_REFRESH_COMPLETE wait respects timeout (default 30). Replace the
prior async=False unit test that only hit the missing-event KeyError path. Document
async and timeout in refresh_modules' docstring and extend the CLI example.
Guard _patch_psutil_pidfd_open_einval(): wait_pid_pidfd_open exists only on newer
psutil (pidfd path). Skip patching when getattr returns None so psutil 5.9.x
environments load conftest without AttributeError during collection.

twangboy commented:
I accidentally merged this and had to revert. Would you mind reopening?
